File: erlang/seminars/philosophers/src/dinner.erl
-module(dinner).

-export([start/1, stop/0]).

start(N) ->
    Seed = 1234,
    Dinner = spawn(fun() -> init(N, Seed) end),
    register(dinner, Dinner).

stop() ->
    dinner ! abort.

init(N, Seed) ->
    C1 = chopstick:start(),    
    C2 = chopstick:start(),
    C3 = chopstick:start(),
    C4 = chopstick:start(),
    C5 = chopstick:start(),
    Ctrl = self(),
    philosopher:start(N, 5, C1, C2, "Arendt", Ctrl, Seed+1),
    philosopher:start(N, 5, C2, C3, "Hypatia", Ctrl, Seed+2),
    philosopher:start(N, 5, C3, C4, "Simone", Ctrl, Seed+3),
    philosopher:start(N, 5, C4, C5, "Elisabeth", Ctrl, Seed+4),
    philosopher:start(N, 5, C1, C5, "Ayn", Ctrl, Seed+5),
    wait(5,[C1,C2,C3,C4,C5]).

wait(0, Chopsticks) ->
    lists:foreach(fun(C) -> chopstick:quit(C) end, Chopsticks);
wait(N, Chopsticks) ->
    receive
	done ->
	    wait(N-1, Chopsticks);
	abort ->
	    erlang:exit(abort)
    end.
    



    
    
    
    
		  

File: exercises/philosophers/src/dinner.ex
defmodule Dinner do

  # Start a dinner.
  def start(n) do
    dinner = spawn(fn -> init(n) end)
    Process.register(dinner, :dinner)
  end

  # Start table at another node
  def start(n, node) do
    dinner = spawn(fn -> init(n, node) end)
    Process.register(dinner, :dinner)
  end  
  
  # Stop the dinner.
  def stop() do
    case Process.whereis(:dinner) do
      nil ->
	:ok
      pid ->
	send(pid, :abort)
    end
  end
  
  defp init(n) do
    c1 = Chopstick.start()
    c2 = Chopstick.start()
    c3 = Chopstick.start()
    c4 = Chopstick.start()
    c5 = Chopstick.start()
    ctrl = self()
    gui = Gai.start([:arendt, :hypatia, :simone, :elisabeth, :ayn])
    Philosopher.start(n, 5, c1, c2, :arendt, ctrl, gui)
    Philosopher.start(n, 5, c2, c3, :hypatia, ctrl, gui)
    Philosopher.start(n, 5, c3, c4, :simone, ctrl, gui)
    Philosopher.start(n, 5, c4, c5, :elisabeth, ctrl, gui)
    Philosopher.start(n, 5, c5, c1, :ayn, ctrl, gui)
    wait(5, [c1, c2, c3, c4, c5])
  end

  defp init(n, node) do
    c1 = Chopstick.start(node)
    c2 = Chopstick.start(node)
    c3 = Chopstick.start(node)
    c4 = Chopstick.start(node)
    c5 = Chopstick.start(node)
    ctrl = self()
    gui = Gai.start([:arendt, :hypatia, :simone, :elisabeth, :ayn])
    Philosopher.start(n, 5, c1, c2, :arendt, ctrl, gui)
    Philosopher.start(n, 5, c2, c3, :hypatia, ctrl, gui)
    Philosopher.start(n, 5, c3, c4, :simone, ctrl, gui)
    Philosopher.start(n, 5, c4, c5, :elisabeth, ctrl, gui)
    Philosopher.start(n, 5, c5, c1, :ayn, ctrl, gui)
    wait(5, [c1, c2, c3, c4, c5])
  end
  
  
  defp wait(0, chopsticks) do
    Enum.each(chopsticks, fn(c) -> Chopstick.quit(c) end)
    Process.unregister(:dinner)
  end

  defp wait(n, chopsticks) do
    receive do
      :done ->
        wait(n - 1, chopsticks)

      :abort ->
	## in order to kill all chopsticks and philosophers
	:io.format("dinner aborted\n")
        Process.exit(self(), :kill)
    end
  end

end
